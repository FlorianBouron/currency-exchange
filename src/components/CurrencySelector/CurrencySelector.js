import React, { Fragment } from "react";
import PropTypes from "prop-types";
import Button from "@material-ui/core/Button";
import Menu from "@material-ui/core/Menu";
import MenuItem from "@material-ui/core/MenuItem";
import KeyboardArrowDown from "@material-ui/icons/KeyboardArrowDown";
import styles from "./CurrencySelector.module.scss";

class CurrencySelector extends React.PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      anchorEl: null
    };
  }

  handleClickButton = event => {
    this.setState({ anchorEl: event.currentTarget });
  };

  handleClickItem = (name, symbol) => {
    const { onChange } = this.props;
    this.setState(
      {
        anchorEl: null
      },
      () => onChange(name, symbol)
    );
  };

  handleCloseMenu = () => {
    this.setState({
      anchorEl: null
    });
  };

  render() {
    const { currentCurrency } = this.props;
    const { anchorEl } = this.state;
    const { wallets } = this.props;
    const menuId = "currency-selector";
    return (
      <Fragment>
        <Button
          size="large"
          aria-owns={anchorEl && menuId}
          aria-haspopup="true"
          onClick={this.handleClickButton}
          classes={{
            label: styles["currency-selector-button"]
          }}
        >
          {currentCurrency}
          <KeyboardArrowDown />
        </Button>
        <Menu
          id={menuId}
          anchorEl={anchorEl}
          open={Boolean(anchorEl)}
          onClose={this.handleCloseMenu}
        >
          {wallets.map(wallet => {
            const { name, symbol, id } = wallet;
            return (
              <MenuItem
                onClick={() => this.handleClickItem(name, symbol)}
                key={id}
              >
                {name}
              </MenuItem>
            );
          })}
        </Menu>
      </Fragment>
    );
  }
}

CurrencySelector.propTypes = {
  currentCurrency: PropTypes.string.isRequired,
  wallets: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
      amount: PropTypes.number.isRequired
    })
  ),
  onChange: PropTypes.func.isRequired
};

CurrencySelector.defaultProps = {
  wallets: []
};

export default CurrencySelector;
