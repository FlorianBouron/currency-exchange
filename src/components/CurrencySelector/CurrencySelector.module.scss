@import "../../constants/utils.scss";

.currency-selector-button {
  font-size: px2rem(23px);
}
