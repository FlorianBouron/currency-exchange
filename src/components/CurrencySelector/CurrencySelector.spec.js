import React from "React";
import { shallow } from "enzyme";
import Button from "@material-ui/core/Button";
import Menu from "@material-ui/core/Menu";
import MenuItem from "@material-ui/core/MenuItem";
import {
  GBP_NAME,
  EUR_NAME,
  USD_NAME,
  GBP_SYMBOL,
  EUR_SYMBOL,
  USD_SYMBOL
} from "../../constants/currencies";
import CurrencySelector from ".";

describe("CurrencySelector component", () => {
  const fakeClickEvent = {
    preventDefault() {}
  };
  const wallets = [
    {
      id: "wallet-id-0",
      name: GBP_NAME,
      amount: 200,
      symbol: GBP_SYMBOL
    },
    {
      id: "wallet-id-1",
      name: EUR_NAME,
      amount: 85,
      symbol: EUR_SYMBOL
    },
    {
      id: "wallet-id-2",
      name: USD_NAME,
      amount: 1000,
      symbol: USD_SYMBOL
    }
  ];
  const basicProps = {
    currentCurrency: "GBP",
    wallets,
    onChange: () => {}
  };

  const component = shallow(<CurrencySelector {...basicProps} />);

  it("has working component", () => {
    expect(component).toBeDefined();
  });

  it("expects component to never be changed", () => {
    expect(component).toMatchSnapshot();
  });

  it("fire the event onClick on Button component", () => {
    expect(component.state("anchorEl")).toBe(null);
    component
      .find(Button)
      .first()
      .simulate("click", fakeClickEvent);
    expect(component.state("anchorEl")).not.toBe(null);
  });

  it("fire the event onClose on the Menu component", () => {
    component
      .find(Button)
      .first()
      .simulate("click", fakeClickEvent);
    expect(component.state("anchorEl")).not.toBe(null);
    component
      .find(Menu)
      .first()
      .simulate("close");
    expect(component.state("anchorEl")).toBe(null);
  });

  it("has working component with no wallets", () => {
    const customProps = { ...basicProps, wallets: [] };
    const component = shallow(<CurrencySelector {...customProps} />);
    expect(component).toBeDefined();
  });

  it("expects component to never be changed with no wallets", () => {
    const customProps = { ...basicProps, wallets: [] };
    const component = shallow(<CurrencySelector {...customProps} />);
    expect(component).toMatchSnapshot();
  });
});
