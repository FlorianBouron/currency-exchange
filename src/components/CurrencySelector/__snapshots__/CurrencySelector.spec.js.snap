// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencySelector component expects component to never be changed 1`] = `
<Fragment>
  <WithStyles(ForwardRef(Button))
    aria-haspopup="true"
    classes={
      Object {
        "label": "currency-selector-button",
      }
    }
    onClick={[Function]}
    size="large"
  >
    GBP
    <[object Object] />
  </WithStyles(ForwardRef(Button))>
  <WithStyles(ForwardRef(Menu))
    anchorEl={null}
    id="currency-selector"
    onClose={[Function]}
    open={false}
  >
    <WithStyles(ForwardRef(MenuItem))
      key="wallet-id-0"
      onClick={[Function]}
    >
      GBP
    </WithStyles(ForwardRef(MenuItem))>
    <WithStyles(ForwardRef(MenuItem))
      key="wallet-id-1"
      onClick={[Function]}
    >
      EUR
    </WithStyles(ForwardRef(MenuItem))>
    <WithStyles(ForwardRef(MenuItem))
      key="wallet-id-2"
      onClick={[Function]}
    >
      USD
    </WithStyles(ForwardRef(MenuItem))>
  </WithStyles(ForwardRef(Menu))>
</Fragment>
`;

exports[`CurrencySelector component expects component to never be changed with no wallets 1`] = `
<Fragment>
  <WithStyles(ForwardRef(Button))
    aria-haspopup="true"
    classes={
      Object {
        "label": "currency-selector-button",
      }
    }
    onClick={[Function]}
    size="large"
  >
    GBP
    <[object Object] />
  </WithStyles(ForwardRef(Button))>
  <WithStyles(ForwardRef(Menu))
    anchorEl={null}
    id="currency-selector"
    onClose={[Function]}
    open={false}
  />
</Fragment>
`;
